### ISSUE TITLE
POST Queries?

### ISSUE THREAD

**Issue by gtamas** [IBM/openapi-to-graphql#193]
A simple question:

Openapi-to-Graphql sees some of the endpoints as mutations. However, these are not really mutations, they are queries.  All these require a POST request to be sent instead of a GET, since they have several parameters and it's easier to transmit those in the request body.

Can I make Openapi-to-Graphql to handle these as Queries?
Do I have to edit the generated GraphQL schema manually to achieve this?

**Comment by gtamas** [comment]
By the way, how exactly OtG map the REST op to GraphQL type? I mean, how does it know if the op is a Query or Mutation? Is this determined by the request method? All GETs are queries and  POST, PUT, DELETE..etc are mutations?

It would be great to find way to use non-GET endpoints as Queries. Using POST method doesn't necessarily  mean the endpoint will mutate data.

**Comment by ErikWittern** [comment]
@gtamas OpenAPI-to-GraphQL does indeed rely on the HTTP method to determine whether to create a query or mutation field. OtG will create query fields only for GET operations, which are both safe and idempotent. POST operations in particular should, according to HTTP, neither be safe nor idempotent (see for example https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST). I think the current behavior is sound, and should remain the default to ensure most users get an expected outcome.

Using POST for a safe, idempotent operation is a bit an anti-pattern, so I am also skeptical of enabling it via configuration. Still, to think this through, how would you propose an option to overwrite the default behavior to look like?

**Comment by gtamas** [comment]
@ErikWittern 

Yes it's true, using POST as query is not recommended, it's actually a sign of bad API design, so therefore the current default behaviour should remain unchanged.

However, we have to accept that people still use such patterns for some reason, so it should be possible to handle these somehow.
 
I think, normally all POST ops should be mutations, but it should be possible to "mark" certain ops, so OtG knows these are actually queries and handles them as such. I see 2 possibilities: 

1. Handle this on the spec level. For example, Operation objects have a "tags" field which is an array of string. So one could add the "query" tag to let OtG (and everybody else :) know this is a query.

2 it should be at least possible to pass some command line args to OtG like:

opeanapi-to-graphql  --qurey someOperation --query someOp2

... where someOperation and someOp2 are  OpenAPI 3 operationIds. 

What do you think?

**Comment by gtamas** [comment]
@ErikWittern 

Actually, the using the tags field is not ideal, but I didn't find any other suitable field in the spec.

However, maybe there is a better way. OtG can already save the generated the GraohQL schema right?. How about adding the ability to load custom GraphQL schema? Then one could simply generate a schema, modify it and then load it in which case OtG could skip the conversion and use the passed schema directly.

Of course this only works if changing the generated schema is easy, like moving the given operation to Query { ... } block from Mutation { ... }. If complex changes are required then this method is not good either.

Maybe this would be cleaner solution.  

Anyway, some kinda solution should be provided to support not only REST but also REST-like APIs. 
Unfortunately, many APIs don't follow the REST spec. This quite common with old, legacy systems which cannot be changed easily (or at all).

What do you think?

**Comment by Alan-Cha** [comment]
@gtamas Interesting proposal! Perhaps instead of using the `tags` field, the user can provide an object that points to a number of operations (or perhaps just the path) similar to the [`customResolvers`](https://github.com/IBM/openapi-to-graphql/tree/master/packages/openapi-to-graphql#options) option. 

On loading and modifying a schema, I think this idea might be outside the scope of OtG. It sounds like it should be a separate tool.

**Comment by ErikWittern** [comment]
I am also not sure about loading modified schemas. OtG currently associates created resolver functions with the data types returned by the corresponding REST operations. Modifying the schema easily breaks that relationship. I currently think this is out of scope.

If certain operations should be treated differently than expected, I would propose to make that an option of OtG. I don't like the idea of extending an OpenAPI description with non-standard things / conventions just for the sake of using that specification with OtG. I.e., I prefer a `--query` and `--mutation` option followed by the operation id that overwrites how OtG decides to treat that operation.

**Comment by gtamas** [comment]
@ErikWittern 
OK I see, cool. Adding the --query / --mutation flags would be great, I also proposed this. Maybe this would be the best solution.

And it should be also possible to do this when OtG is used as a lib. Something  like @Alan-Cha  proposes would be nice.

**Comment by gtamas** [comment]
@wtrocki @ErikWittern @Alan-Cha 

Actually, this sounds like a really cool idea but unfortunately it doesn't seem to work.

Whenever I mutate the schema this way, all queries return something like this:

```
{
  "data": {
    "someType": null
  }
}
```

In fact, this happens even I don't return any value in leave() / enter() methods,  in which case nodes won't be changed at all, so the schema should be unchanged.

I guess, this is because buildASTSchema() returns something with no resolvers and thus it cannot be used to execute queries, as stated here:

https://graphql.org/graphql-js/utilities/#buildastschema

Do I understand this correctly? Or am I doing something wrong? It seems to me that the schema created by OtG's createGraphQlSchema() cannot be modified with graphqlJS?

**Comment by gtamas** [comment]
Any . other ideas to try? Anyone?

Is there any way to modify (post-process) the generated GQL Schema **and** make it work with OtG?
I really don't want to fork this and change the code.

It should be possible to make changes to the schema without losing functionality... As @wtrocki  pointed out, there are several use cases. A solution like the one he proposed would give us lots of flexibility.

**Comment by Alan-Cha** [comment]
@gtamas I'm so sorry. I let this slip by! I've been super busy and it's been difficult to keep up with everything.

Actually, in regards to forking and changing the code. For us, that is **highly encouraged** if those changes can lead to some kind of PR. That is the reason why we open-sourced, so we could work with a community to improve OtG. 

I think we need to implement some kind of [`customResolvers`](https://github.com/IBM/openapi-to-graphql/tree/master/packages/openapi-to-graphql#options) option like I proposed earlier. 

If you want to begin that effort, I would be happy to help you!

**Comment by gtamas** [comment]
@Alan-Cha 

OK thanks. I'm also pretty busy right now, but will check the code. :) Maybe we could fork it.

But just for clarification: Should @wtrocki's solution work or not?

```
import { printSchema, buildASTSchema, parse } from 'graphql';
import { createGraphQlSchema } from 'openapi-to-graphql';
...
  let { schema: graphQLSchema, report: graphQLReport } = await createGraphQlSchema(openAPISchema, options);

graphQLSchema = buildASTSchema(parse(printSchema(graphQLSchema)))
```
Executing this produces a schema which is valid, but none of the queries return any meaningful results. Without the  last line, however, it works fine.

So should the schema generated by buildASTSchema() be compatible with OtG? 
Another question: are the resolvers part of the GraphQL schema OtG creates, or a resolver map has to be added separately?

**Comment by gtamas** [comment]
Regarding adding an object like customResolvers, that would be cool! 

However, there could other similar issues as well. 

For example, how about adding custom links?

Use case: I'm working here with several REST APIs whose OAS does not have link objects, so OtG cannot generate links.
So maybe adding an option that allows one to pass link definitions (something like customResolvers) would be also useful.

That's why I tried @wtrocki's idea. A mechanism like this would be able to solve several such issues.

**Comment by gtamas** [comment]
@Alan-Cha 

OK so I'm not familiar with the code yet. Could you help me get started? First I would look into adding this object we discussed (custom operation mapping). But some basic info on how the relevant part of the code works would be appreciated. :)

Thanks in advance!

**Comment by Alan-Cha** [comment]
@gtamas I'm actually not too sure if @wtrocki's solution will work for this case. It can modify the schema itself but I don't think it can modify the resolvers. You would have to generate a new set of resolvers some other way. 

To answer your question, the resolvers are contained within the schema (GraphQLSchemaObject) object that OtG creates. 

***

I see what you mean about your use case but is it not possible to update them so that they do have link objects? I'm guessing they may be auto-generated? On one hand, I would like to direct users towards a single option, an option that is naturally provided by OAS, instead of externalizing it, but on the other hand, this might make OtG work better with auto-generated OASs. I would love to get the opinion of @ErikWittern and @wtrocki again if that's possible!

***

I am very happy to give you some pointers!

We preprocess (extracting the data) the OAS and create the `data` object [here](https://github.com/IBM/openapi-to-graphql/blob/44385871286982574b96bc60a9e505ef2a2b7570/packages/openapi-to-graphql/src/index.ts#L209).

We create the GraphQL schema by iterating through all the `operations` stored in `data` [here](https://github.com/IBM/openapi-to-graphql/blob/44385871286982574b96bc60a9e505ef2a2b7570/packages/openapi-to-graphql/src/index.ts#L221). 

There is an if statement [here](https://github.com/IBM/openapi-to-graphql/blob/44385871286982574b96bc60a9e505ef2a2b7570/packages/openapi-to-graphql/src/index.ts#L239) that will determine whether an operation will be turned into a `Query` field or a `Mutation` field. 

I think this should just be a matter of adding some additional logic to that if statement. We just need to check if the operation is mentioned in the option, maybe just call it `selectQueryOrMutationField` for now, and if it is, respond appropriately. 

We will need to add the new option [here](https://github.com/IBM/openapi-to-graphql/blob/44385871286982574b96bc60a9e505ef2a2b7570/packages/openapi-to-graphql/src/types/options.ts#L29) and [here](https://github.com/IBM/openapi-to-graphql/blob/44385871286982574b96bc60a9e505ef2a2b7570/packages/openapi-to-graphql/src/types/options.ts#L166). We process the options [here](https://github.com/IBM/openapi-to-graphql/blob/44385871286982574b96bc60a9e505ef2a2b7570/packages/openapi-to-graphql/src/index.ts#L75) and [here](https://github.com/IBM/openapi-to-graphql/blob/44385871286982574b96bc60a9e505ef2a2b7570/packages/openapi-to-graphql/src/index.ts#L177). 

I don't think it's that much work. Along with that, we will need to update the documentation and add a test case which I am happy to help you with when we reach that point.

**Comment by Alan-Cha** [comment]
@gtamas The `selectQueryOrMutationField` option should have a similar shape as `customResolvers`. You can see how the latter is defined [here](https://github.com/IBM/openapi-to-graphql/blob/44385871286982574b96bc60a9e505ef2a2b7570/packages/openapi-to-graphql/src/types/options.ts#L117) but instead of containing a `ResolveFunction`, I think it should point to an enum (which will we also need to define) containing just `Query` and `Mutation`.

**Comment by gtamas** [comment]
@Alan-Cha 

Thanks for the detailed info! I tried to follow your instructions and just opened a PR.
Now it is possible to make any operation a query or mutation.

https://github.com/IBM/openapi-to-graphql/pull/279

Please check it out! I'm not entirely sure if it's all OK, since this is my first PR here and I'm not that familiar with the code yet :)

Tests and docs are yet to be added. If you could provide some help with that, it would be great. :)

Thanks in advance

**Comment by Alan-Cha** [comment]
@gtamas This is a great start and thanks for taking a deep breath and diving in! I will try to give you some input later today! Honestly, this is fantastic work. We really appreciate your help!

**Comment by gtamas** [comment]
@Alan-Cha 

OK thank you very much! :)

**Comment by gtamas** [comment]
@Alan-Cha 

Hi, any news about this :)?

**Comment by Alan-Cha** [comment]
@gtamas I made some changes! Take a look and let me know what you think!

**Comment by Alan-Cha** [comment]
Closed with PR #282 

Thanks @gtamas for your contributions!!!

**Comment by gtamas** [comment]
@Alan-Cha 

Thank you very much for merging and all the help! I think this feature will be very useful.
